### PR TITLE
NV center files

### PR DESCRIPTION
**Pull request by yomach**

### CODE REVIEW

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/README.md
Comment (danielLQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/README.md:1]: change to 
# Single NV center in a confocal setup

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py
Comment (danielLQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py:17]: meas_len = laser_len_cycles * 4 + 1000  # total measurement length (ns)

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter.py
Comment (danielLQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter.py:11]: single_integration_time_ns = int(500e3)  # 500us

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter_all.py
Comment (danielLQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter_all.py:11]: single_integration_time_ns = int(500e3)  # 500us

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py
Comment (danielLQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py:13]: maybe say that a_min < d/2?

Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py:13]: Not sure what you mean here?

Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py:13]: a_min = 0.9
da = 0.1
a_max = 1

I get, 0.9, 1.0

Don't see the issue...
The problem is if you choose a_min = 0.91.
In this case you'll also get 1.1, which is higher than a_max.
However, it will also be included in the scan, so it is ok...

The only real issue is if you pick a_min=0.95, and then 1.05 will not be included in numpy `arange`, but **might** be included in our scan.

Comment (danielLQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py:13]: you're right, the relation is actually more annoying, if you choose 0< a_min< d/2 +n*da . But it is still there.
for instance
a_min=0.02/0.12/0.22...
da=0.1
a_max=1
then you also get 1.02

Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py:13]: Yes, but this is the same as you would get using numpy `arange`.
With this I am at peace...

Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py:13]: I mean that numpy `arange` would give the same results are our `for_` loop, so I am ok with that.

Comment (danielLQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/power_rabi.py:13]: I just noticed you change the for_ loop condition for a fixed variable, My bad, ignore

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py
Comment (TheoLaudatQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py:80]: Oh right sorry in this one yes, but not in Power Rabi or Time Rabi

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays_python_histogram.py
Comment (TheoLaudatQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays_python_histogram.py:34]: The config and the other files use "NV" instead of "nv" and "const" instead of "mw"

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py
Comment (TheoLaudatQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py:45]: It seems that this program is deriving the histogram as opposed to the one called calibrate_delays_python_histogram.py

**Review by TheoLaudatQM — CHANGES_REQUESTED** [review]
Maybe as a general comment I would suggest to add a short docstring at the beginning of each program that explains what it does in a line or two. I think this is particularly important for these programs that the whole team is going to use eventually. I am thinking about something like this but it is open:
"""
rabi_amp_freq.py: template for acquiring the 2D (pulse amplitude & frequency sweeps) Rabi oscillations
Author: Theo Laudat - Quantum Machines
Created: 02/05/2022
"""

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/hahn_echo.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/hahn_echo.py:41]: Measuring once with X and once with -X (it's a pi flip).
Maybe you're correct and it's better to put it in Macro, but not right now :)

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/hahn_echo.py
Comment (TheoLaudatQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/hahn_echo.py:52]: Ok, I understand!

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/hahn_echo.py
Comment (TheoLaudatQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/hahn_echo.py:41]: Yes sorry, alright I'll had it to the list for future improvement :)

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/cw_odmr.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/cw_odmr.py:31]: Hmm. This function could be nice regardless.
Actually, one which allows you to easily scan measure pulses length (with integration duration) would be awesome for other cases.

But I'm not sure I would use it here...
The only issue is when doing sequences with short timings (and specifically with arbitrary waveforms), here no gaps would be created...
I agree that it's not 100% according to our best practices...

Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/cw_odmr.py:31]: I will say that in our best practice we say that we recommend to use the duration and amp only for calibration scripts, and this is a calibration script :)

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter.py:1]: Yes, but the save_all is very resource heavy, we wouldn't want to use it all the time...
In most cases we would just use the normal one

Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter.py:1]: Maybe this one can be removed, not sure

**Comment by yomach** [comment]
> Maybe as a general comment I would suggest to add a short docstring at the beginning of each program that explains what it does in a line or two. I think this is particularly important for these programs that the whole team is going to use eventually. I am thinking about something like this but it is open: """ rabi_amp_freq.py: template for acquiring the 2D (pulse amplitude & frequency sweeps) Rabi oscillations Author: Theo Laudat - Quantum Machines Created: 02/05/2022 """

We can add that.
The original idea was to have in the readme file all of the scripts, according to the order in which they normally should be run (specifically useful for calibration scripts), with a short 1-2 lines of explanation.
We need to add that ASAP.
I personally prefer to have the scripts themselves a bit cleaner, but I don't mind adding it on the top.
I wouldn't add a date/author though, because these things change and different people do changes... You can always look in the git to see who did what (Git Blame!)

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays_python_histogram.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays_python_histogram.py:17]: It is there, 4 lines above...

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays.py:45]: yes

**Comment by TheoLaudatQM** [comment]
> > Maybe as a general comment I would suggest to add a short docstring at the beginning of each program that explains what it does in a line or two. I think this is particularly important for these programs that the whole team is going to use eventually. I am thinking about something like this but it is open: """ rabi_amp_freq.py: template for acquiring the 2D (pulse amplitude & frequency sweeps) Rabi oscillations Author: Theo Laudat - Quantum Machines Created: 02/05/2022 """
> 
> We can add that. The original idea was to have in the readme file all of the scripts, according to the order in which they normally should be run (specifically useful for calibration scripts), with a short 1-2 lines of explanation. We need to add that ASAP. I personally prefer to have the scripts themselves a bit cleaner, but I don't mind adding it on the top. I wouldn't add a date/author though, because these things change and different people do changes... You can always look in the git to see who did what (Git Blame!)

I understand your point and I think that the readme is also necessary, but it only takes a few lines at the beginning allows the user to directly know what the program does without having to go to the github page... 
This is just my opinion and it can be a topic to discuss with the others to see what the best would be :)

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays_python_histogram.py
Comment (TheoLaudatQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays_python_histogram.py:17]: No I meant in the derivation of `meas_len` to be sure that you are measuring long enough. If for instance you change `initial_delay_cycles` to 5000 for some reason, then you will not measure properly, unless I misunderstood something

Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/calibrate_delays_python_histogram.py:17]: Ahh, I see, good point.

**Comment by yomach** [comment]
> > > Maybe as a general comment I would suggest to add a short docstring at the beginning of each program that explains what it does in a line or two. I think this is particularly important for these programs that the whole team is going to use eventually. I am thinking about something like this but it is open: """ rabi_amp_freq.py: template for acquiring the 2D (pulse amplitude & frequency sweeps) Rabi oscillations Author: Theo Laudat - Quantum Machines Created: 02/05/2022 """
> > 
> > 
> > We can add that. The original idea was to have in the readme file all of the scripts, according to the order in which they normally should be run (specifically useful for calibration scripts), with a short 1-2 lines of explanation. We need to add that ASAP. I personally prefer to have the scripts themselves a bit cleaner, but I don't mind adding it on the top. I wouldn't add a date/author though, because these things change and different people do changes... You can always look in the git to see who did what (Git Blame!)
> 
> I understand your point and I think that the readme is also necessary, but it only takes a few lines at the beginning allows the user to directly know what the program does without having to go to the github page... This is just my opinion and it can be a topic to discuss with the others to see what the best would be :)

As I said, I don't mind adding it to the top of the files (But without the author/date).
I'll do that tomorrow for these files, and we need to do it for the SC soon

**Comment by TheoLaudatQM** [comment]
> > > > Maybe as a general comment I would suggest to add a short docstring at the beginning of each program that explains what it does in a line or two. I think this is particularly important for these programs that the whole team is going to use eventually. I am thinking about something like this but it is open: """ rabi_amp_freq.py: template for acquiring the 2D (pulse amplitude & frequency sweeps) Rabi oscillations Author: Theo Laudat - Quantum Machines Created: 02/05/2022 """
> > > 
> > > 
> > > We can add that. The original idea was to have in the readme file all of the scripts, according to the order in which they normally should be run (specifically useful for calibration scripts), with a short 1-2 lines of explanation. We need to add that ASAP. I personally prefer to have the scripts themselves a bit cleaner, but I don't mind adding it on the top. I wouldn't add a date/author though, because these things change and different people do changes... You can always look in the git to see who did what (Git Blame!)
> > 
> > 
> > I understand your point and I think that the readme is also necessary, but it only takes a few lines at the beginning allows the user to directly know what the program does without having to go to the github page... This is just my opinion and it can be a topic to discuss with the others to see what the best would be :)
> 
> As I said, I don't mind adding it to the top of the files (But without the author/date). I'll do that tomorrow for these files, and we need to do it for the SC soon

Perfect I am good with this !

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter_all.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/counter_all.py]: The inner loop is for having numbers (Notice we only save the total counts). We actually ignore the statistics (might want to add them).
The outer loop is for it to run continuously. The counter is usually turned on for a long period of time while playing with the optics.

**Comment by yomach** [comment]
@TheoLaudatQM - Please take a look at the readme file and the descriptions. If it's good I'll also add it to all of the files.

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py
Comment (TheoLaudatQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py:103]: Printing the remaining time is indeed very nice and the guys in Finland also implemented something similar, so I think it will soon be a new feature in the py_qua_tools!

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py
Comment (TheoLaudatQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py:55]: It seems that you measure T1 for the two states anyway so I don't understand the purpose of start_from_one... Why would the order matter?

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/README.md
Comment (TheoLaudatQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/README.md:20]: Perfect!

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py
Comment (TheoLaudatQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py:79]: I am waiting for Guy comments, but as soon as it is merged and ready we'll replace al of these by the new fetching tool

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py
Comment (TheoLaudatQM) [inline Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py:114]: Just out of curiosity, why do you always measure along X and -X, is it standard for NV centers ?

Comment (danielLQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py:114]: In room temperature NV measurements you get photons from both the |0> and |+-1> states. So comparing the two projections is a good indication for the state contrast compared to a Rabi measurement. In addition, this is a good way to differentiate between features (like an echo revival, or a decoherence peak),  and noise.

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py:79]: Yes, definitely!

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py:55]: It's not the same.
In room-temperature NV experiments, the decay is to a mixed state (not to |0>)
So both |0> and |1> decay.
For both cases, we measure both the population at |0> and at |1> for the same reasons Daniel wrote below...
So we can both initialize to |0> or |1> and measure at |0> or |1>!

Now let me complicate things and say that you can also initialize |-1> and measure at |-1>!
But these measurements are more rare

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py
Comment (yomach) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py:103]: :) 
Also see what I did for plotting last figure.
And what I did to get double X-axis (specifically for the CW ODMR, which is similar to spectroscopy)

py-qua-tools here we come!

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py
Comment (TheoLaudatQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py:55]: Ok I see, thanks I'll need to read about how NVs work !

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py
Comment (TheoLaudatQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py:114]: I see, thanks !

File: Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py
Comment (TheoLaudatQM) [reply Quantum-Control-Applications/NV-Centers/Confocal-Setup/T1.py:103]: Yes it very cool, there are a lot of new stuff to add :)

**Review by TheoLaudatQM — APPROVED** [review]

**Pull request merged**
